server/middleware/oidc/index.js

```diff
@@ -0,0 +1,123 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+const OpenIDClient = require('openid-client');
+const passport = require('koa-passport');
+const session = require('koa-session');
+const allowUrl = ['/login', '/logout', '/oauth2/redirect'];
+
+let settings = {};
+
+try {
+  settings = require('../../config/oidc');
+} catch (e) {
+  console.log('OIDC configuration file not found, using ENV variables');
+}
+
+const clientID = settings.clientID || process.env.OPENID_CLIENT_ID;
+const clientSecret = settings.clientSecret || process.env.OPENID_CLIENT_SECRET;
+const callbackURL = settings.callbackURL || process.env.OPENID_CALLBACK_URL;
+const discoverURL = settings.discoverURL || process.env.OPENID_DISCOVER_URL;
+const scope = settings.scope || process.env.OPENID_SCOPE || 'openid';
+
+const middleware = async function(ctx, next) {
+  if (allowUrl.includes(ctx.path)) {
+    return next();
+  }
+
+  if (ctx.isUnauthenticated()) {
+    return ctx.redirect('/login');
+  }
+
+  // logout user to remove session with old token
+  if (
+    ctx.state.user.exp !== undefined &&
+    ctx.state.user.exp < Date.now() / 1000
+  ) {
+    ctx.logout();
+
+    return ctx.redirect('/login');
+  }
+
+  ctx.authTokenHeaders = ctx.authTokenHeaders || {};
+  ctx.authTokenHeaders['cadence-authorization'] = ctx.state.user.accessToken;
+
+  await next();
+};
+
+const setupAuth = async function(app, router) {
+  app.keys = ['secret'];
+  app.use(session(app));
+  app.use(passport.initialize());
+  app.use(passport.session());
+
+  router.get('/login', passport.authenticate('oidc'));
+  router.get('/logout', async ctx => {
+    ctx.logout();
+    ctx.redirect('/');
+  });
+  router.get(
+    '/oauth2/redirect',
+    passport.authenticate('oidc', {
+      successReturnToOrRedirect: '/',
+      failureRedirect: '/login',
+    })
+  );
+  passport.serializeUser((user, done) => done(null, user));
+  passport.deserializeUser((user, done) => done(null, user));
+
+  const discovered = await OpenIDClient.Issuer.discover(discoverURL);
+  const openIdClient = new discovered.Client({
+    client_id: clientID,
+    client_secret: clientSecret,
+  });
+
+  const openIDStrategy = new OpenIDClient.Strategy(
+    {
+      client: openIdClient,
+      params: {
+        redirect_uri: callbackURL,
+        scope: scope,
+      },
+      passReqToCallback: false,
+      sessionKey: 'koa.sess',
```

Comment:
This can have a better name

Reply:
This is dictated by https://github.com/koajs/session/blob/98630aaf6587b199969490f8503641554aeee917/index.js#L63

But ideally, this should be inherited from upstream class somehow. Need to check

Reply:
@mantas-sidlauskas from what i see, this value is just the default value if you didn't pass a key. But my expectation is that you can pass one. 

Maybe you have an issue with it because your first had the default value and changing it didn't change the created session.

Reply:
actually, it's not needed. removed this key altogether